client.go

```diff
@@ -192,12 +201,20 @@ func (c *Client) exchange(m *Msg, a string) (r *Msg, rtt time.Duration, err erro
 	}
 
 	co.TsigSecret = c.TsigSecret
-	co.SetWriteDeadline(time.Now().Add(c.writeTimeout()))
+	writeDeadline := deadline
```

Comment:
Perhaps a helper function instead of these 4 additional lines of code multiple times?

Something like
```
func deadlineOrTimeout(deadline time.Time, timeout time.Duration) time.Time {
  if deadline.IsZero() {
    return time.Now().Add(timeout)
  }
  return deadline
}
```

Then in the main function use something like
```
co.SetWriteDeadline(deadlineOrTimeout(deadline, c.writeTimeout())
...
co.SetReadDeadline(deadlineOrTimeout(deadline, c.readTimeout())`
...
```